L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoClusteringConfig_SA.cc

```diff
@@ -0,0 +1,145 @@
+      stepLatency_({{Input, 0},
+                    {Dist0, 0},
+                    {Dist1, 0},
+                    {Dist2, 0},
+                    {Dist3, 0},
+                    {Dist4, 0},
+                    {Dist5, 0},
+                    {TcToHc, 0},
+                    {Hist, 0},
+                    {Smearing1D, 0},
+                    {NormArea, 0},
+                    {Smearing2D, 0},
+                    {Maxima1D, 0},
+                    {Maxima2D, 0},
+                    {CalcAverage, 0},
+                    {Clusterizer, 0},
+                    {TriggerCellToCluster, 0},
+                    {ClusterSum, 0}}),
+      depths_(),
+      triggerLayers_(),
+      layerWeights_E_(),
+      layerWeights_E_EM_(),
+      layerWeights_E_EM_core_(),
+      layerWeights_E_H_early_(),
+      correction_(),  // 0b011111111111111111
+      saturation_()   // (2 ** 19) - 1
+{}
+
+void ClusterAlgoConfig::setStepLatencies(const std::vector<unsigned int> latencies) {
+  // Add check that stepLatency is at least same size as latencies
+  // But not as cms.exception
+  for (unsigned int iStep = 0; iStep < latencies.size(); ++iStep) {
+    stepLatency_.at(Step(iStep)) = latencies.at(iStep);
+  }
+}
+
+unsigned int ClusterAlgoConfig::getLatencyUpToAndIncluding(const Step step) {
+  unsigned int latency = 0;
+  for (int iStep = 0; iStep <= step; ++iStep)
+    latency += getStepLatency(Step(iStep));
+  return latency;
+}
+
+void ClusterAlgoConfig::initializeLUTs() {
+  initializeSmearingKernelConstants(cRows_, rOverZHistOffset_, rOverZBinSize_);
+  initializeThresholdMaximaConstants(cRows_, thresholdMaximaParam_a_, thresholdMaximaParam_b_, thresholdMaximaParam_c_);
+  initializeCosLUT();
+}
+
+void ClusterAlgoConfig::initializeSmearingKernelConstants(unsigned int bins, unsigned int offset, unsigned int height) {
+  const unsigned int lWidth0 = offset + (0.5 * height);
+  const unsigned int lTarget = int((maxBinsSmearing1D_ + 0.5) * lWidth0 - 0.5);
+  for (unsigned int iBin = 0; iBin < bins; ++iBin) {
+    unsigned int lCentre = lWidth0 + (height * iBin);
+    const unsigned int lBins = int(round(1.0 * lTarget / lCentre));
+
+    kernelWidths_.push_back(lBins);
+
+    lCentre *= lBins;
+
+    const unsigned int lRatio = int(round(1.0 * lTarget / lCentre * pow(2, nBitsAreaNormLUT_)));
+
+    areaNormalizations_.push_back(lRatio);
+  }
+}
+
+void ClusterAlgoConfig::initializeThresholdMaximaConstants(unsigned int bins, unsigned int a, unsigned int b, int c) {
+  for (unsigned int iBin = 0; iBin < bins; ++iBin) {
+    int threshold = a + b * iBin + c * iBin * iBin;
+    thresholdMaximaConstants_.push_back(threshold);
+  }
+}
+
+void ClusterAlgoConfig::initializeCosLUT() {
+  for (unsigned int iBin = 0; iBin < nBinsCosLUT_ + 1; ++iBin) {
+    unsigned int cosBin = round(pow(2, nBitsCosLUT_) * (1 - cos(iBin * M_PI / phiNValues_)));
+    cosLUT_.push_back(cosBin);
+  }
+}
+
+void ClusterAlgoConfig::printConfiguration() {
```

Comment:
This was mainly for debugging early on in development.  I am happy to delete it.  It may still be useful to dump something like this via the MessageLogger or to file, but not in this way.